src/model_repository_manager.cc

```diff
@@ -768,24 +822,63 @@ ModelRepositoryManager::LoadUnloadModels(
   }
   std::set<ModelIdentifier> deleted_dependents;
 
-  // Update dependency graph and load
-  UpdateDependencyGraph(
-      added, deleted, modified,
+  // Make a copy of the dependency graph for rollback if there is a collision
+  // between the models to be loaded/unloaded by this thread with another
+  // thread actively loading/unloading models. The global map referenced by the
+  // dependency graph is managed by this object, which is the responsibility of
+  // this object to make a copy.
+  std::unordered_map<std::string, std::set<ModelIdentifier>> global_map(
+      global_map_);
+  DependencyGraph dependency_graph(dependency_graph_, &global_map);
+
+  // Update the dependency graph independently for the possibility of rollback.
+  auto affected_models = UpdateDependencyGraph(
+      &dependency_graph, infos, added, deleted, modified,
       unload_dependents ? &deleted_dependents : nullptr);
 
+  // Check for collision in affected models and lock those models
+  auto conflict_model = dependency_graph.LockUnlockNodes(
+      affected_models, true /* to_lock */,
+      &dependency_graph_ /* prev_dependency_graph */);
+  if (conflict_model) {
+    // A collision is found. Since info is only written to local copy, so it is
+    // safe to rollback by simply returning the conflict infomation.
+    if (no_parallel_conflict != nullptr) {
+      *no_parallel_conflict = false;
+      return Status::Success;
+    }
+    return Status(
+        Status::Code::INTERNAL,
+        "a related model '" + conflict_model->str() +
+            "' to a load/unload request is currently loading or unloading");
+  }
+
+  // At this point, it is committed to load/unload the models, so write the
+  // local copy into this object. The state of the models cannot be modified by
+  // another thread while they are locked. Thus, it is safe to release the
+  // mutex while actively loading/unloading models, but the mutex must be
+  // re-held when reading/updating any model info to avoid any corruption.
+  infos_.swap(infos);
+  global_map_.swap(global_map);
+  dependency_graph_.swap(dependency_graph);
+
```

Comment:
You don't really need the lock afterward, which should save you from all the reverse locking. Idea is that by this point, the structure has been committed, which implies the "node" under change doesn't interfere with other (/ previous) ongoing model operation and there shouldn't be concurrent access on any data structure.

Then you only need to acquire the lock again once the whole operation is completed and unmarked the flag in dependency graph.

Reply:
Then I will bring back the writeback step, because the loaded info needs to be written back to the master records held by the manager after loading completes.

The steps will be: acquire (make a copy), update, commit (set the copy as ground truth), and writeback.

Reply:
Update commit: [Refactor parallel load logic](https://github.com/triton-inference-server/core/pull/165/commits/0fa4e296b613806d0bd7de77775f09e51d866d21)